src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs

```diff
@@ -2,45 +2,138 @@
+        private readonly Dictionary<string, AssemblyLoadContext> _loadContextByDirectory = new Dictionary<string, AssemblyLoadContext>();
 
         protected override Assembly LoadFromPathImpl(string fullPath)
         {
-            //.NET Native doesn't support AssemblyLoadContext.GetLoadContext. 
-            // Initializing the _loadContext in the .ctor would cause
-            // .NET Native builds to fail because the .ctor is called. 
-            // However, LoadFromPathImpl is never called in .NET Native, so 
-            // we do a lazy initialization here to make .NET Native builds happy.
-            if (_loadContext == null)
+            AssemblyLoadContext? loadContext;
+
+            var fullDirectoryPath = Path.GetDirectoryName(fullPath) ?? throw new ArgumentException(message: null, paramName: nameof(fullPath));
+            lock (_guard)
             {
-                AssemblyLoadContext loadContext = AssemblyLoadContext.GetLoadContext(typeof(DefaultAnalyzerAssemblyLoader).GetTypeInfo().Assembly);
+                if (!_loadContextByDirectory.TryGetValue(fullDirectoryPath, out loadContext))
+                {
+                    loadContext = new DirectoryLoadContext(fullDirectoryPath, this);
+                    _loadContextByDirectory[fullDirectoryPath] = loadContext;
+                }
+            }
+
+            var name = AssemblyName.GetAssemblyName(fullPath);
+            return loadContext.LoadFromAssemblyName(name);
+        }
 
-                if (System.Threading.Interlocked.CompareExchange(ref _loadContext, loadContext, null) == null)
+        private bool ShouldLoadInAnalyzerContext(string? assemblySimpleName, string fullPath)
+        {
+            foreach (var compilerAssemblySimpleName in CompilerAssemblySimpleNames)
+            {
+                if (string.Equals(compilerAssemblySimpleName, assemblySimpleName, StringComparison.OrdinalIgnoreCase))
                 {
-                    _loadContext.Resolving += (context, name) =>
-                    {
-                        Debug.Assert(ReferenceEquals(context, _loadContext));
-                        return Load(name.FullName);
-                    };
+                    return false;
                 }
             }
 
-            return LoadImpl(fullPath);
+            return IsKnownDependencyLocation(fullPath);
         }
 
-        protected virtual Assembly LoadImpl(string fullPath) => _loadContext.LoadFromAssemblyPath(fullPath);
+        private sealed class DirectoryLoadContext : AssemblyLoadContext
+        {
+            private readonly string _directory;
+            private readonly DefaultAnalyzerAssemblyLoader _loader;
+
+            public DirectoryLoadContext(string directory, DefaultAnalyzerAssemblyLoader loader)
+            {
+                _directory = directory;
+                _loader = loader;
+            }
+
+            protected override Assembly? Load(AssemblyName assemblyName)
```

Comment:
There are two places in this method where we return `null`. The implication of returning `null` is that the runtime will then move to the compiler ALC and attempt to load `AssemblyName` there. In the two places we return `null` though that doesn't seem to be the behavior we want. The load should only succeed in this ALC and given that it failed we shouldn't defer to compiler ALC we should likely fail. Think we should just throw an exception to prevent that.

Reply:
Ah! Yeah, I forgot about this. It seems like if the dependency belongs to the compiler, we should return null, but if the dependency is simply unknown then we throw to be sure that no other ALC will try to resolve it.